include/zephyr/senss/senss.h

```diff
@@ -0,0 +1,486 @@
+ */
+int senss_set_interval(
+		senss_sensor_handle_t handle,
+		uint32_t value);
+
+/**
+ * @brief Set current data change sensitivity of given client handle.
+ *
+ * Since each sensor type may have multiple data fields for it's value, this
+ * API support set separated sensitivity for each data field, or global
+ * sensitivity for all data fields.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param index The value fields index to be set, -1 for all fields (global).
+ *
+ * @param value The value to be set.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+int senss_set_sensitivity(
+		senss_sensor_handle_t handle,
+		int index, uint32_t value);
+
+/**
+ * @brief Get current data change sensitivity of given client handle.
+ *
+ * Since each sensor type may have multiple data fields for it's value, this
+ * API support get separated sensitivity for each data field, or global
+ * sensitivity for all data fields.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param index The value fields index to be set, -1 for all fields (global).
+ *
+ * @param value The data buffer to receive value.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+int senss_get_sensitivity(
+		senss_sensor_handle_t handle,
+		int index, uint32_t *value);
+
+/**
+ * @brief Read a data sample
+ *
+ * This API will trigger target sensor read a sample each time.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param buf The data buffer to receive sensor data sample.
+ *
+ * @param size The buffer size in bytes.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_read_sample(
+		senss_sensor_handle_t handle,
+		void *buf, int size);
+
+/**
+ * @brief Trigger a sensor to do self calibration
+ *
+ * Some sensor support auto self calibration.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+int senss_self_calibration(
+		senss_sensor_handle_t handle);
+
+/**
+ * @brief Sensor data event receive callback.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param buf The data buffer with sensor data.
+ *
+ * @param size The buffer size in bytes.
+ *
+ * @param param The user private parameter.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+typedef int (*senss_data_event_t)(
+		senss_sensor_handle_t handle,
+		void *buf, int size, void *param);
+
+/**
+ * @brief Register sensor data event receive callback.
+ *
+ * User can use this API to register a data event receive callback function.
+ *
+ * @param handle The sensor instance handle.
+ *
+ * @param callback The data event receive callback function.
+ *
+ * @param param The user private parameter.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_register_data_event_callback(
```

Comment:
Since the register information size (handle, callback pointer, private parameter pointer) is not big, I think another method is internally pre-allocate memory (a static array etc) for max possible clients, also we can use kconfig to config the max count. 

The benefit of this approach is can keep the API simple, intuit and easy use, no need user define a structure and may contain internal data structure.

Reply:
The down side here is runtime errors which might be harder to find. I personally prefer to have these structs allocated by the client, but we should discuss with the greater community

Reply:
Client allocated structure will expose some internal implement data fields,  such as the `next` pointer etc, and makes the API not easy use for client and not flexible if need  dynamic memory allocation in future. 

I prefer a simple and straightforward API, just like current version,  the implementations can have internal static memory allocation scheme (use pre-defined static array etc) or dynamic memory allocation scheme based on platform (some platform may have rich memory resource) and requirements. We can implement the static memory allocation scheme at first, but in future we can implement the dynamic memory allocation scheme also and use `Kconfig` to config the schemes, all of these is just impact the internal implementations, but not impacts to API and Applications.

Agree, we should discuss with the greater community. :)